`ultralytics 8.3.127` New Visual Similarity Search Solution

RizwanMunawar opened the pull request:
## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://github.com/ultralytics/actions)<sub>

### 🌟 Summary
A new semantic image search solution has been added to Ultralytics, enabling users to search and retrieve images using natural language queries powered by OpenAI CLIP and Meta FAISS, with a user-friendly web interface. 🖼️🔍

### 📊 Key Changes
- **New Solution: Semantic Image Search**
  - Introduced `VisualAISearch` and `SearchApp` classes for semantic image search using CLIP and FAISS.
  - Added a Flask-powered web app for interactive, natural language image search.
  - Created comprehensive documentation and a user guide for the new feature.
  - Added a stylish HTML template for the search interface.
- **Documentation & Navigation Updates**
  - Added guides and reference docs for similarity search.
  - Updated navigation menus to include the new solution.
- **Testing & Dependencies**
  - Added tests for the similarity search solution.
  - Included Flask as a dependency for solutions.
- **Core Enhancements**
  - Updated solution configuration to support image directory selection.
  - Improved solution class structure to support the new feature.

### 🎯 Purpose & Impact
- **Empowers Users**: Anyone can now search large image collections using plain English, without needing labels or manual annotation.
- **Zero-Shot Search**: Instantly find relevant images with flexible, human-like queries—no retraining required.
- **Easy Integration**: The solution is accessible via a simple Python API or a ready-to-use web app.
- **Scalable & Fast**: FAISS ensures efficient search even with thousands of images, making it suitable for both small and large datasets.
- **Customizable**: The web interface can be easily tailored to fit different projects or branding needs.
- **Broader Use Cases**: Ideal for creative tools, archives, product search, and more—opening new possibilities for Ultralytics users.

🚀 This update makes advanced AI-powered image search accessible to everyone in the Ultralytics ecosystem!

RizwanMunawar committed:
update

RizwanMunawar committed:
update

RizwanMunawar committed:
update

RizwanMunawar committed:
update

RizwanMunawar committed:
Merge 06d294343257629aa16591dda48d48427ee4474c into e700646ea265326aad8822aa6db2a7d4fabacf85

UltralyticsAssistant committed:
Auto-update Ultralytics Docs Reference by https://ultralytics.com/actions

UltralyticsAssistant commented:
👋 Hello @RizwanMunawar, thank you for submitting an `ultralytics/ultralytics` 🚀 pull request! This is an automated response to help streamline the review process. An Ultralytics engineer will also assist you soon.

Please review the following checklist to ensure your PR can be merged smoothly:

- ✅ **Define a Purpose**: Clearly explain the purpose of your fix or feature in your PR description, and link to any [relevant issues](https://github.com/ultralytics/ultralytics/issues). Make sure your commit messages are clear and follow project conventions.
- ✅ **Synchronize with Source**: Confirm your PR is up-to-date with the `ultralytics/ultralytics` `main` branch. If it's behind, update it by clicking the 'Update branch' button or running `git pull` and `git merge main` locally.
- ✅ **Ensure CI Checks Pass**: Ensure all Ultralytics [Continuous Integration (CI)](https://docs.ultralytics.com/help/CI/) checks are passing. If any checks fail, please address them.
- ✅ **Update Documentation**: Update the relevant [documentation](https://docs.ultralytics.com/) for any new or changed features, as you have started in `visual_similarity_search.md`.
- ✅ **Add Tests**: If applicable, include or update tests that cover your changes, and confirm all tests are passing.
- ✅ **Sign the CLA**: If this is your first contribution to Ultralytics, please sign our [Contributor License Agreement (CLA)](https://docs.ultralytics.com/help/CLA/) by commenting "I have read the CLA Document and I sign the CLA" in this PR.
- ✅ **Minimize Changes**: Limit your changes to only what is necessary for this feature addition. _"It is not daily increase but daily decrease, hack away the unessential. The closer to the source, the less wastage there is."_ — Bruce Lee

For complete contributing guidelines, refer to our [Contributing Guide](https://docs.ultralytics.com/help/contributing/). If you have any questions, please leave a comment here.

Thank you for helping make Ultralytics better! 🖼️✨

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

RizwanMunawar committed:
update

RizwanMunawar committed:
Merge fb1a76d84d6e02b576f623a1a489d1608328b57a into fba1b3f513da0aafc2d821eb4475f4ac5d8a238c

UltralyticsAssistant committed:
Auto-update Ultralytics Docs Reference by https://ultralytics.com/actions

RizwanMunawar committed:
update

RizwanMunawar committed:
Merge 1cfe04d86630d8de057551de7f691be452464b50 into fba1b3f513da0aafc2d821eb4475f4ac5d8a238c

UltralyticsAssistant committed:
Auto-update Ultralytics Docs Reference by https://ultralytics.com/actions

RizwanMunawar committed:
update

RizwanMunawar committed:
update

RizwanMunawar committed:
Merge branch 'main' into semantic-search

RizwanMunawar committed:
Merge d889802e212007679a26caa9130df9ab2af40b90 into 1a212e990f28986d6319e57b09f0bd6cfe4e33df

UltralyticsAssistant committed:
Auto-update Ultralytics Docs Reference by https://ultralytics.com/actions

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

RizwanMunawar committed:
update

RizwanMunawar committed:
update

RizwanMunawar committed:
update

RizwanMunawar committed:
update

RizwanMunawar committed:
update

RizwanMunawar committed:
update

RizwanMunawar committed:
update

RizwanMunawar committed:
Merge branch 'main' into semantic-search

RizwanMunawar committed:
Merge 0f287b98aa74032679c5219ae4e5bfec8cddae54 into e1d742500fbfc24a08c4eb664d0111eaffc8827a

UltralyticsAssistant committed:
Auto-update Ultralytics Docs Reference by https://ultralytics.com/actions

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

RizwanMunawar committed:
update

RizwanMunawar committed:
update

RizwanMunawar committed:
Merge e58f61e991477639772caa8ed6084d3b200c7057 into e1d742500fbfc24a08c4eb664d0111eaffc8827a

UltralyticsAssistant committed:
Auto-update Ultralytics Docs Reference by https://ultralytics.com/actions

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

RizwanMunawar committed:
update

RizwanMunawar committed:
update

RizwanMunawar committed:
Merge 01840b167edc9771cce484e56945405c1be2c489 into e1d742500fbfc24a08c4eb664d0111eaffc8827a

UltralyticsAssistant committed:
Auto-update Ultralytics Docs Reference by https://ultralytics.com/actions

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

RizwanMunawar committed:
update

RizwanMunawar committed:
update

RizwanMunawar committed:
Merge 8ac96562a22a335d7e731b3fe4adfa77fdc3ed87 into e1d742500fbfc24a08c4eb664d0111eaffc8827a

UltralyticsAssistant committed:
Auto-update Ultralytics Docs Reference by https://ultralytics.com/actions

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

RizwanMunawar committed:
update

RizwanMunawar committed:
Merge branch 'semantic-search' of https://github.com/ultralytics/ultralytics into semantic-search

RizwanMunawar committed:
Merge e3aea049c4cbb670418c526a4874b1102d5429c2 into e1d742500fbfc24a08c4eb664d0111eaffc8827a

UltralyticsAssistant committed:
Auto-update Ultralytics Docs by https://ultralytics.com/actions

RizwanMunawar committed:
update

RizwanMunawar committed:
Merge branch 'semantic-search' of https://github.com/ultralytics/ultralytics into semantic-search

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

RizwanMunawar committed:
update

RizwanMunawar committed:
Merge branch 'semantic-search' of https://github.com/ultralytics/ultralytics into semantic-search

RizwanMunawar committed:
update

RizwanMunawar committed:
update

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

RizwanMunawar committed:
update

RizwanMunawar committed:
Merge branch 'semantic-search' of https://github.com/ultralytics/ultralytics into semantic-search

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

RizwanMunawar committed:
Update pyproject.toml

Signed-off-by: Muhammad Rizwan Munawar <muhammadrizwanmunawar123@gmail.com>

RizwanMunawar committed:
Add test for similarity search

Signed-off-by: Muhammad Rizwan Munawar <muhammadrizwanmunawar123@gmail.com>

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

RizwanMunawar committed:
Update test_solutions.py

Signed-off-by: Muhammad Rizwan Munawar <muhammadrizwanmunawar123@gmail.com>

RizwanMunawar commented:
@glenn-jocher Hi, can you please review this PR? It's the new solution, `similarity search`. The webpage preview is here:

![2](https://github.com/user-attachments/assets/a2628478-9633-496b-af64-15df7efae934)

Here's the code for testing.


### Similarity search basic

```python
from ultralytics import solutions

searcher = solutions.VisualAISearch(
    # data = "path/to/img/directory" # Optional, build search engine with your own images
    device="cuda"  # configure the device for processing, i.e "cpu" or "cuda"
)

results = searcher.search("a dog sitting on a bench")
```

**Output**
```
# Ranked Results:
#     - 000000546829.jpg | Similarity: 0.3269
#     - 000000549220.jpg | Similarity: 0.2899
#     - 000000517069.jpg | Similarity: 0.2761
#     - 000000029393.jpg | Similarity: 0.2742
#     - 000000534270.jpg | Similarity: 0.2680
```


### Similarity search with a web application

```python
from ultralytics import solutions

app = solutions.SearchApp(
    # data = "path/to/img/directory" # Optional, build search engine with your own images
    device="cpu"  # configure the device for processing, i.e "cpu" or "cuda"
)

app.run(debug=False)  # You can also use `debug=True` argument for testing
```

Thanks

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

RizwanMunawar commented:
@glenn-jocher, all tests are passing. Also I manually run the Scheduled CI and everything is passing here as well: https://github.com/ultralytics/ultralytics/actions/runs/14774906016

Thanks

UltralyticsAssistant committed:
Merge branch 'main' into semantic-search

glenn-jocher reviewed (commented):
@RizwanMunawar looks really good! Should the template index.html have a more descriptive filename, or if it needs to be called index.html maybe placed in a more descriptive subdirectory?

RizwanMunawar commented:
@glenn-jocher yes, we can rename the file, maybe `similarity-search.html`?

glenn-jocher commented:
@RizwanMunawar yeah that makes sense. Check to make sure it's not required to be called index.html though, I know some web formats (like MkDocs in certain modes) have a mode where instead of HTML filenames the files are called index.html and the slug gets created from the directory name.

Like this: https://www.mkdocs.org/user-guide/configuration/#use_directory_urls

RizwanMunawar committed:
Merge branch 'main' into semantic-search

RizwanMunawar committed:
Adjust template filename and align with BaseSolution

RizwanMunawar committed:
Merge 982cc5647f5a9861e9fc0c5078224b913e056b98 into c4252487cfd2d559a3f9e79090127d8cae877739

UltralyticsAssistant committed:
Auto-update Ultralytics Docs Reference by https://ultralytics.com/actions

RizwanMunawar committed:
update solutions.py

RizwanMunawar committed:
update test_solutions.py

glenn-jocher commented:
@RizwanMunawar can you resolve the conflicts here?

glenn-jocher committed:
Update __init__.py

Signed-off-by: Glenn Jocher <glenn.jocher@ultralytics.com>

RizwanMunawar commented:
@glenn-jocher Ok, I updated the template to `similarity-search.html`, since it's a Flask-based application, we can adjust the template filename, the important thing is that it should be inside the `templates` folder, that's the trick here.

Besides that, I adjusted the `device` and `data` arguments inside the config file to align with the existing solutions we have. The general CI is passing here, I think. I just triggered the [Scheduled CI](https://github.com/ultralytics/ultralytics/actions/runs/14822124377) and [Docker tests](https://github.com/ultralytics/ultralytics/actions/runs/14822122542), let's wait for everything to be passed and then we can merge. In case, anything we need to update anything, just let me know. Thanks

Edit: Let me resolve the conflicts. I think these are due to the new PR we just merged.

RizwanMunawar committed:
update solutions.py

RizwanMunawar committed:
Merge branch 'semantic-search' of https://github.com/ultralytics/ultralytics into semantic-search

RizwanMunawar committed:
Merge branch 'main' into semantic-search

RizwanMunawar committed:
update solutions.py

glenn-jocher commented:
@ultralytics/run-ci

glenn-jocher commented:
@RizwanMunawar PR merged, nice work!!

UltralyticsAssistant commented:
🎉 Fantastic work @RizwanMunawar and @glenn-jocher—your semantic image search solution is officially part of Ultralytics! This is a true leap forward for our community. As Albert Einstein said, “Imagination is more important than knowledge. For knowledge is limited, whereas imagination embraces the entire world.” Your contribution embodies this spirit—empowering users to discover images with natural language, sparking new creative possibilities across the Ultralytics ecosystem. Thank you both for your innovation, dedication, and for making advanced AI search accessible to all! 🚀

Laughing-q commented:
@RizwanMunawar Hey nice work on this! I saw we are using another `open_clip` package in this PR, can we reuse the `clip` package as we used for YOLO-World models? https://github.com/ultralytics/ultralytics/blob/149dd26afe280e9893db4e94f8bd7e8a9d1bb845/ultralytics/nn/text_model.py#L12-L16
Also is it possible to set `moible_clip` model(that we are using for YOLOE models) as an option here? I think mobile_clip would definitely get faster speed.

RizwanMunawar commented:
@Laughing-q, great suggestion! I'm not certain about the accuracy of `mobileclip`, but we can include it as an option. I'll update the code to use it by default, while still allowing users to switch back to clip via the model argument. I’ll try to work on this later this week.
